x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/hash/Hash.java

```diff
@@ -0,0 +1,115 @@
+import org.elasticsearch.compute.ann.Evaluator;
+import org.elasticsearch.compute.ann.Fixed;
+import org.elasticsearch.compute.operator.EvalOperator;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.expression.function.scalar.EsqlScalarFunction;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+
+import java.io.IOException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HexFormat;
+import java.util.List;
+
+public class Hash extends EsqlScalarFunction {
+
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(Expression.class, "Hash", Hash::new);
+
+    private final Expression alg;
+    private final Expression input;
+
+    @FunctionInfo(returnType = "keyword", description = "Computes the hash of the input using the supplied algorithm.")
+    public Hash(
+        Source source,
+        @Param(name = "alg", type = { "keyword", "text" }, description = "Hash algorithm to use.") Expression alg,
+        @Param(name = "input", type = { "keyword", "text" }, description = "Input to hash.") Expression input
+    ) {
+        super(source, List.of(alg, input));
+        this.alg = alg;
+        this.input = input;
+    }
+
+    private Hash(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(alg);
+        out.writeNamedWriteable(input);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataType.KEYWORD;
+    }
+
+    @Override
+    public boolean foldable() {
+        return alg.foldable() && input.foldable();
+    }
+
+    @Evaluator(warnExceptions = NoSuchAlgorithmException.class)
+    static BytesRef process(BytesRef alg, BytesRef input) throws NoSuchAlgorithmException {
+        return hash(MessageDigest.getInstance(alg.utf8ToString()), input);
+    }
+
+    @Evaluator(extraName = "Constant")
+    static BytesRef processConstant(@Fixed(build = true) MessageDigest alg, BytesRef input) {
+        return hash(alg, input);
+    }
+
+    private static BytesRef hash(MessageDigest alg, BytesRef input) {
+        alg.update(input.bytes, input.offset, input.length);
+        var result = alg.digest();
+        return new BytesRef(HexFormat.of().formatHex(result));
```

Comment:
I feel like we have three or four hand rolled hex encoders. I might add a flavor that encodes directly into a `BytesRef` into `MessageDigests`. Something like

```
    static BytesRef process(@Fixed(includeInToString = false) BreakingBytesRefBuilder scratch, BytesRef alg, BytesRef input) throws NoSuchAlgorithmException {
        ...
        alg.update(input.bytes, input.offset, input.length);
       byte[] digest = alg.digest();
       scratch.clear();
       scratch.grow(digest.length * 2);
       MessageDigests.toHexUtf8Bytes(scratch.bytes(), digest);
       return scratch.bytesRefView();
    }
```

Something like that. It'd skip a copy or two. And it'd allow us to reuse the allocations from last time.